### PR TITLE
Tree: Fix siblings on cursor.up()

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

More cursor fixes in `up`, and more test coverage for it.

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts [hunk 1]
```diff
@@ -398,18 +398,25 @@ class Cursor implements ITreeSubscriptionCursor {
         return { result: TreeNavigationResult.NotFound, moved: 0 };
     }
     up(): TreeNavigationResult {
-        if (this.parentStack.pop() === undefined) {
-            // We are at the root, so return NotFound without making any changes to the state.
+        const length = this.parentStack.length;
+        assert(this.indexStack.length === length, "Unexpected indexStack.length");
+        assert(this.keyStack.length === length - 1, "Unexpected keyStack.length");
+
+        // Already at the root
```
Comment (jenn-le) [inline packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:405]: It took me a little time to understand why we're already at the root even though there's a parent, could you add a quick explanation here?

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts [hunk 2]
```diff
@@ -398,18 +398,25 @@ class Cursor implements ITreeSubscriptionCursor {
...
             return TreeNavigationResult.NotFound;
         }
+
+        assert(length > 1, "Unexpected parentStack.length");
+        this.parentStack.pop();
         this.indexStack.pop();
         this.keyStack.pop();
         // TODO: maybe compute siblings lazily or store in stack? Store instead of keyStack?
-        if (this.parentStack.length === 0) {
+        if (length === 2) {
```
Comment (jenn-le) [inline packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:415]: nit: This also required a double take for me. I think either adding a comment or assigning meaning to this value would help.

**Review by jenn-le — APPROVED** [review]

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts [hunk 3]
```diff
@@ -398,18 +398,25 @@ class Cursor implements ITreeSubscriptionCursor {
...
             // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
             this.siblings = this.forest.getRoot(this.root!);
         } else {
-            const newParent = this.parentStack[this.parentStack.length - 1];
+            const newParent = this.parentStack[this.parentStack.length - 2];
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts]: I agree. Renamed.

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:416]: comments added.

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts (hunk 1 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts]: comments added.

**Review by DLehenbauer — APPROVED** [review]
I haven't yet absorbed TextCursor, but pre-approving on faith that you know what you're doing and it's important to keep code flowing.

**Pull request merged**